x-pack/solutions/security/plugins/cloud_security_posture/server/routes/graph/fetch_graph.ts

```diff
@@ -172,64 +177,130 @@ const buildEsqlQuery = ({
   const SECURITY_ALERTS_PARTIAL_IDENTIFIER = '.alerts-security.alerts-';
   const enrichPolicyName = getEnrichPolicyId(spaceId);
 
+  const actorFieldsWhereClause = GRAPH_ACTOR_ENTITY_FIELDS.map(
+    (field) => `${field} IS NOT NULL`
+  ).join(' OR ');
+
+  const actorFieldsCoalesce = GRAPH_ACTOR_ENTITY_FIELDS.join(',\n    ');
+
   const query = `FROM ${indexPatterns
     .filter((indexPattern) => indexPattern.length > 0)
     .join(',')} METADATA _id, _index
-| WHERE event.action IS NOT NULL AND actor.entity.id IS NOT NULL
-| MV_EXPAND actor.entity.id
-| MV_EXPAND target.entity.id
+| WHERE event.action IS NOT NULL AND (${actorFieldsWhereClause})
+| EVAL actorEntityId = COALESCE(
+    ${actorFieldsCoalesce}
+  )
+| EVAL targetEntityId = user.target.entity.id
+| EVAL targetEntityId = CASE(
+    host.target.entity.id IS NULL,
+    targetEntityId,
+    CASE(
+      targetEntityId IS NULL,
+      host.target.entity.id,
+      MV_DEDUPE(MV_APPEND(targetEntityId, host.target.entity.id))
+    )
+  )
+| EVAL targetEntityId = CASE(
+    service.target.entity.id IS NULL,
+    targetEntityId,
+    CASE(
+      targetEntityId IS NULL,
+      service.target.entity.id,
+      MV_DEDUPE(MV_APPEND(targetEntityId, service.target.entity.id))
+    )
+  )
+| EVAL targetEntityId = CASE(
+    entity.target.id IS NULL,
+    targetEntityId,
+    CASE(
+      targetEntityId IS NULL,
+      entity.target.id,
+      MV_DEDUPE(MV_APPEND(targetEntityId, entity.target.id))
+    )
+  )
+| MV_EXPAND actorEntityId
+| MV_EXPAND targetEntityId
+| EVAL actorEntityFieldHint = CASE(
+    user.entity.id IS NOT NULL, "user",
+    host.entity.id IS NOT NULL, "host",
+    service.entity.id IS NOT NULL, "service",
+    entity.id IS NOT NULL, "entity",
+    ""
+  )
+| EVAL targetEntityFieldHint = CASE(
+    MV_CONTAINS(user.target.entity.id, targetEntityId), "user",
+    MV_CONTAINS(host.target.entity.id, targetEntityId), "host",
+    MV_CONTAINS(service.target.entity.id, targetEntityId), "service",
+    MV_CONTAINS(entity.target.id, targetEntityId), "entity",
+    ""
+)
```

Comment:
Why don't we evaluate `targetEntityFieldHint` using the same method as in `actorEntityFieldHint`? 

Unless I'm missing some detail, it should be possible and more performant than calling `MV_CONTAINS`

```suggestion
| EVAL actorEntityFieldHint = CASE(
    user.entity.id IS NOT NULL, "user",
    host.entity.id IS NOT NULL, "host",
    service.entity.id IS NOT NULL, "service",
    entity.id IS NOT NULL, "entity",
    ""
  )
| EVAL targetEntityFieldHint = CASE(
    user.target.entity.id IS NOT NULL, "user",
    host.target.entity.id IS NOT NULL, "host",
    service.target.entity.id IS NOT NULL, "service",
    entity.target.id IS NOT NULL, "entity",
    ""
)
```